checker/src/main/java/alluxio/checker/MapReduceIntegrationChecker.java

```diff
@@ -0,0 +1,285 @@
+        }
+      }
+      mMergedAddress.set(mergedAddress);
+      context.write(key, mMergedAddress);
+    }
+  }
+
+  /**
+   * Creates the HDFS filesystem to generate input and output files.
+   *
+   * @param conf Hadoop configuration
+   */
+  private void createHdfsFilesystem(Configuration conf) throws Exception {
+    // Inits HDFS File System Object
+    String systemUserName = System.getProperty("user.name");
+    System.setProperty("HADOOP_USER_NAME", systemUserName);
+    System.setProperty("hadoop.home.dir", "/");
+    mFileSystem = FileSystem.get(URI.create(conf.get("fs.defaultFS")), conf);
+
+    String stringPath = "/user/" + systemUserName + "/";
+    Path userPath = new Path(stringPath);
+
+    // Creates user folder if not exists
+    if (!mFileSystem.exists(userPath)) {
+      mFileSystem.mkdirs(userPath);
+      LOG.info("User Path " + stringPath + " created.");
+    }
+    mInputFilePath = new Path(userPath + "/MapReduceInputFile");
+    mOutputFilePath = new Path(userPath + "/MapReduceOutputFile");
+  }
+
+  /**
+   * Creates the MapReduce input file and delete previous output file if exists.
+   */
+  private void createInputFile(int inputSplits) throws Exception {
+    if (!mFileSystem.exists(new Path(mInputFilePath + "/1.txt"))) {
+      LOG.info("Begin Write MapReduce input file into HDFS");
+      for (int i = 1; i <= inputSplits; i++) {
+        FSDataOutputStream inputFileStream = mFileSystem
+            .create(new Path(mInputFilePath + "/" + i + ".txt"));
+        inputFileStream.writeByte(1);
+        inputFileStream.close();
+      }
+      LOG.info("End Write MapReduce input file into HDFS");
+    }
+
+    if (mFileSystem.exists(mOutputFilePath)) {
+      mFileSystem.delete(mOutputFilePath, true);
+    }
+  }
+
+  /**
+   * @return result Status summarization
+   */
+  private Status generateReport() throws Exception {
+    // Reads all the part-r-* files in MapReduceOutPutFile folder
+    mFileSystem.delete(new Path(mOutputFilePath + "/_SUCCESS"), true);
+    FileStatus[] outputFileStatus = mFileSystem.listStatus(mOutputFilePath);
+
+    boolean canFindClass = true;
+    boolean canFindFS = true;
+
+    StringBuilder cannotFindClassBuilder = new StringBuilder();
+    StringBuilder cannotFindFSBuilder = new StringBuilder();
+    StringBuilder successBuilder = new StringBuilder();
+
+    for (int i = 0; i < outputFileStatus.length; i++) {
+      Path curOutputFilePath = outputFileStatus[i].getPath();
+      try (BufferedReader curOutputFileReader = new BufferedReader(
+          new InputStreamReader(mFileSystem.open(curOutputFilePath)))) {
+        String nextLine = "";
+        while ((nextLine = curOutputFileReader.readLine()) != null) {
+          Status curOutputStatus = Status.valueOf(nextLine
+              .substring(0, nextLine.indexOf(' ')).trim());
+          String curOutputAddresses = nextLine.substring(nextLine.indexOf(' ') + 1).trim();
+          switch (curOutputStatus) {
```

Comment:
the logic of going from a `[(status, hostname), (status, hostname)]` report to writing a file is the same between the Spark and Hadoop checkers. Could we refactor to avoid the duplication?

Reply:
The logic is the same, but the output of Spark(Tuple2<Status, String>) and MapReduce (lines in part-r-*) are so different and the output messages are different. To use a common utils method may not help to make codes more clear.

Reply:
if we replace "Spark executors" and "Hadoop nodes" with just "Nodes", the meaning will still be clear and the outputs will be the same.

it simplifies the logic if we break the code into two phases: first parsing and then analysis. It isn't hard to parse either of the results into a `Map<Status, List<String>>`, and from then on everything is the same.